include/emulator/ppu.h

```diff
@@ -0,0 +1,77 @@
+#pragma once
+
+#include "constants.h"
+#include "nes_ppu_memory_accessor.h"
+
+class PPU {
+    private:
+        Byte pattern_table_[8 * 1024] = {0}; // static ch-rom
+        Byte name_table_[4 * 1024]= {0}; // dynamic vram
+        Byte palette_[32]= {0}; // color palette
+        Byte oam_[64]= {0}; // object attribute memory
+
+        struct PPUCtrl
+        {
+            union
+            {
+                Byte data; // 0bVPHBSINN
+                struct {
+                    Byte nametable0 : 1; //
+                    Byte nametable1 : 1; // ^ base nametable address
+                    Byte increment : 1; // vram increment per cpu read/write of PPUDATA
+                                        // if 0, increment 1 else increment 32
+                    Byte sprite_addr : 1; // sprite pattern table address
+                    Byte background_addr : 1; // background pattern table address
+                    Byte sprite_size : 1; // sprite size
+                    Byte master_slave_select : 1; // PPU master/slave select
+                    Byte vblank : 1; // generate NMI at start of vblank
+                } flags;
+            };
+        } reg_ctrl_; // PPUCTRL $2000 > Write
+
+        struct PPUMask
+        {
+            union
+            {
+                Byte data; // 0bBGRsBMmG
+                struct {
+                    Byte greyscale : 1; // produce greyscale display
+                    Byte background_left : 1; // show background in leftmost 8 pixels of screen
+                    Byte sprite_left : 1; // show sprites in leftmost 8 pixels of screen
+                    Byte background : 1; // show background
+                    Byte sprite : 1; // show sprite
+                    Byte red : 1; // emphasize red
+                    Byte green : 1; // emphasize green
+                    Byte blue : 1; // emphasize blue
+                } flags;
+            };
+        } reg_mask_; // PPUMASK $2001 > Write
+
+        struct PPUStatus
+        {
+            union
+            {
+                Byte data; // 0bVSO-----
+                struct {
+                    Byte : 5; // not used
+                    Byte overflow :1; // sprite overflow
+                    Byte sprite_hit : 1; // sprite 0 hit
+                    Byte vblank_started : 1; // vblank has started
+                } flags;
+            };
+        } reg_status_; // PPUMASK $2001 > Write
+
+        Word temp_address_ = 0; // used with ppuaddr
+        Word vram_address_ = 0;
+        Byte latch_ = 0; // indicates firts or second write in ppuaddr
+        Byte interal_buffer_ = 0; // holds discarded value on first read from ppudata
+        NESPPUMemoryAccessor *memory_;
+
+
+    public:
+        PPU(NESPPUMemoryAccessor *memory);
+        PPU(NESPPUMemoryAccessor *memory, Byte ctrl_data);
+        void WriteToAddrReg(Byte data);
```

Comment:
Where are we calling these?  It feels weird to have to inject the PPU as a direct dependency in the CPU memory space.  Perhaps we introduce events into our memory accessors and allow classes to subscribe to writes to certain areas?  I'd think that make isolating for testing easier.

Reply:
Thanks for the feedback! My thought was that when the cpu instructs the memory accessor to read to or write from cpu memory, if the location is within 2000-2007, it would call a particular ppu register read/write function. These ppu register functions would then call the ppu memory accessor functions. In my mind, this is how the ppu/cpu communicate together in the nes hardware. I am open to other ideas, and I'm not sure what you mean exactly by introducing events into the memory accessors and allowing classes to subscribe. If you could elaborate on that or provide an example, I might have a better way to move forward with this branch.